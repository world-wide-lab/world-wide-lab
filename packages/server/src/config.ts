import 'dotenv/config';

function getValueFromEnv(key: string): string | undefined {
  const value = process.env[key]
  // This catches undefined, null and ""
  if (!value) {
    return undefined
  } else {
    return value
  }
}

function getStringFromEnv(key: string): string {
  const value = getValueFromEnv(key)
  if (value === undefined) {
    throw new Error(`${key} must not be empty!`);
  }
  return value
}

function getBooleanFromEnv(key: string, defaultValue: boolean = false): boolean {
  const value = getValueFromEnv(key)
  if (value === undefined) {
    return defaultValue
  }
  const stringValue = value.toLowerCase()
  if (stringValue === "true") {
    return true
  } else if (stringValue === "false") {
    return false
  } else {
    throw new Error(`Invalid value for ${key}: ${value}. Only "true" and "false" are supported.`)
  }
}

const config = {
  root: getValueFromEnv("ROOT") || "http://localhost",
  port: getValueFromEnv("PORT") || 8787,

  version: process.env.npm_package_version as string,

  logging: {
    dir: getValueFromEnv("LOGGING_DIR") || "logs",
    // Change to 'verbose' to log SQL queries
    consoleLevel: getValueFromEnv("LOGGING_LEVEL_CONSOLE") || "info",
    http: getBooleanFromEnv("LOGGING_HTTP", true),
  },

  admin: {
    enabled: getBooleanFromEnv("ADMIN_UI", true),
    auth: {
      enabled: getBooleanFromEnv("USE_AUTHENTICATION", true),
      sessionSecret: getValueFromEnv("ADMIN_AUTH_SESSION_SECRET"),
      default_admin_credentials: {
        email: getValueFromEnv("ADMIN_AUTH_DEFAULT_EMAIL"),
        password: getValueFromEnv("ADMIN_AUTH_DEFAULT_PASSWORD")
      },
    },
  },

  apiDocs: {
    enabled: getBooleanFromEnv("API_DOCS", true),
  },

  api: {
    apiKey: getValueFromEnv("DEFAULT_API_KEY"),
  },

  database: {
    url: getStringFromEnv("DATABASE_URL"),
    generateExampleData: getBooleanFromEnv("GENERATE_EXAMPLE_DATA", true),
  }
}

// Validate configuration
if (config.admin.enabled && config.admin.auth.enabled) {
  if (
    !config.admin.auth.sessionSecret ||
    !config.admin.auth.default_admin_credentials.email ||
    !config.admin.auth.default_admin_credentials.password
  ) {
    throw new Error(`When authentication for Admin UI is enabled, credentials and session secret have to be set as well.`)
  }
}

export default config