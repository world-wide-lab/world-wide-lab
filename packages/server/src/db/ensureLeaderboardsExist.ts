import type { Sequelize } from "sequelize";

async function ensureLeaderboardsExist(
  sequelize: Sequelize,
  leaderboardIds: string[],
) {
  await Promise.all(
    leaderboardIds.map((leaderboardId) =>
      ensureLeaderboardExists(sequelize, leaderboardId),
    ),
  );
}

async function ensureLeaderboardExists(
  sequelize: Sequelize,
  leaderboardId: string,
) {
  const existingLeaderboardCount = (
    await sequelize.models.Leaderboard.findAndCountAll({
      where: {
        leaderboardId,
      },
    })
  ).count;

  if (existingLeaderboardCount > 0) {
    // Leaderboard already exists
    return;
  }

  // Create leaderboard
  await sequelize.models.Leaderboard.create({
    leaderboardId,
  });
}

export { ensureLeaderboardsExist, ensureLeaderboardExists };
