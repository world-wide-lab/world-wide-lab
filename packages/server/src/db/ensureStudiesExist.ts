import type { Sequelize } from "sequelize";

async function ensureStudiesExist(sequelize: Sequelize, studyIds: string[]) {
  await Promise.all(
    studyIds.map((studyId) => ensureStudyExists(sequelize, studyId)),
  );
}

async function ensureStudyExists(sequelize: Sequelize, studyId: string) {
  const existingStudyCount = (
    await sequelize.models.Study.findAndCountAll({
      where: {
        studyId,
      },
    })
  ).count;

  if (existingStudyCount > 0) {
    // Study already exists
    return;
  }

  // Create study
  await sequelize.models.Study.create({
    studyId,
  });
}

export { ensureStudiesExist, ensureStudyExists };
