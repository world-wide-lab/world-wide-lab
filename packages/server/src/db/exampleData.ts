import { Sequelize } from "sequelize";
import { logger } from "../logger";

async function generateExampleData(
  sequelize: Sequelize,
  studyId: string = "example",
) {
  const exampleStudyCount = (
    await sequelize.models.Study.findAndCountAll({
      where: {
        studyId,
      },
    })
  ).count;

  if (exampleStudyCount > 0) {
    // Data has already been generated
    return;
  }
  logger.info("Generating Example Data");
  await sequelize.models.Study.create({
    studyId,
  });

  async function generateParticipantData(data: {
    privateInfo: Object;
    sessions: Array<number>;
    createdAt?: Date;
    finished_after_sessions?: number;
  }) {
    const { privateInfo, sessions, createdAt, finished_after_sessions } = data;
    const participant = await sequelize.models.Participant.create({
      privateInfo,
      createdAt,
    });
    for (let sessionIndex = 0; sessionIndex < sessions.length; sessionIndex++) {
      const n_responses = sessions[sessionIndex];
      const session = await sequelize.models.Session.create({
        // @ts-ignore
        participantId: participant.participantId,
        studyId,
        createdAt,
        finished:
          finished_after_sessions === undefined
            ? false
            : n_responses >= finished_after_sessions,
      });
      const responses = [];
      for (let index = 1; index <= n_responses; index++) {
        responses.push({
          // @ts-ignore
          sessionId: session.sessionId,
          name: `trial-${index}`,
          payload: {
            response: `Response #${index}`,
          },
        });
      }
      await sequelize.models.Response.bulkCreate(responses);
    }
  }

  function daysAgo(daysAgo: number): Date {
    const date = new Date();
    date.setDate(date.getDate() - daysAgo);
    return date;
  }

  await generateParticipantData({
    privateInfo: {
      description: "Example User #1",
    },
    sessions: [3],
    finished_after_sessions: 3,
    createdAt: daysAgo(4),
  });
  await generateParticipantData({
    privateInfo: {
      description: "Example User #2",
    },
    sessions: [3],
    finished_after_sessions: 3,
    createdAt: daysAgo(3),
  });
  await generateParticipantData({
    privateInfo: {
      description: "Example User #3",
    },
    sessions: [3, 3, 1, 3],
    finished_after_sessions: 3,
    createdAt: daysAgo(2),
  });
  await generateParticipantData({
    privateInfo: {
      description: "Example User #4",
    },
    sessions: [3, 1],
    finished_after_sessions: 3,
    createdAt: daysAgo(1),
  });
}

export default generateExampleData;
