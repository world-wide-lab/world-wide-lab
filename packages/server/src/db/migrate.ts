import pathLib from "node:path";
import { SequelizeStorage, Umzug } from "umzug";

import process from "node:process";
import { fileURLToPath } from "node:url";
import { logger } from "../logger.js";
import { getDirectory } from "../util.js";
import sequelize from "./index.js";

const dirname = getDirectory(import.meta.url);

const dialect = sequelize.getDialect();

const umzug = new Umzug({
  // Support both common and dialect-specific migrations
  migrations: {
    glob: [`migrations/*.@(common|${dialect}).@(js|ts)`, { cwd: dirname }],
    resolve: ({ name, path, context }) => {
      const getMigration = async () => import(path as string);
      const nameWithoutExtension = pathLib.parse(path as string).name;

      // adjust the parameters Umzug will
      // pass to migration methods when called
      return {
        // Remove extension of migration names (to avoid double applying between js and ts)
        name: nameWithoutExtension,
        up: async () => {
          const migration = await getMigration();
          await migration.up({ context });
        },
        down: async () => {
          const migration = await getMigration();
          await migration.down({ context });
        },
      };
    },
  },
  context: sequelize.getQueryInterface(),
  storage: new SequelizeStorage({
    sequelize,
    modelName: "InternalMigrations",
    tableName: "wwl_internal_migrations",
  }),
  // Map winston onto the umzug internal logger version
  logger: {
    info: (message) => logger.info,
    warn: (message) => logger.warn,
    error: (message) => logger.error,
    debug: (message) => logger.info,
  },
});

async function up() {
  const pending = await umzug.pending();
  if (pending.length > 0) {
    const migrationNames: Array<string> = Object.values(pending).map(
      (migration) => migration.name,
    );
    logger.info("Applying pending migrations: ", migrationNames);
  } else {
    logger.info("No pending migrations");
  }

  // Checks migrations and run them if they are not already applied. To keep
  // track of the executed migrations, a table (and sequelize model) called SequelizeMeta
  // will be automatically created (if it doesn't exist already) and parsed.
  await umzug.up();
}

let latestMigration: string;
async function getLatestMigration(includeSuffix: boolean): Promise<string> {
  if (latestMigration === undefined) {
    const executed = await umzug.executed();
    latestMigration = executed[executed.length - 1].name;
  }
  if (includeSuffix) {
    return latestMigration;
  }
  return latestMigration.substring(0, latestMigration.lastIndexOf("."));
}

// Detect whether the file is called directly
// TODO: Find a cleaner solution for this and determine whether we even need this
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  // Can be called via e.g.
  // node dist/db/migrate.js create --name migration-name.ts --skip-verify
  // after npm run build, or the following before building
  // npx node --loader ts-node/esm src/db/migrate.ts create --name migration-name.ts --skip-verify
  umzug.runAsCLI();
}

// export the type helper exposed by umzug, which will have the `context` argument typed correctly
export type Migration = typeof umzug._types.migration;

export { umzug, up, getLatestMigration };
