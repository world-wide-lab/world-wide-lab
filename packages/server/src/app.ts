import path from "path";
import cors from "cors";
import express, { Request, Response } from "express";
import helmet from "helmet";

import { admin, adminRouter } from "./admin";
import api from "./api";
import apiDocs from "./api-docs";
import { routerProtectedWithoutAuthentication } from "./api/protected";
import config from "./config";
import { logger } from "./logger";

const app = express();

// app.use(helmet());
app.use(cors());
app.use(express.json());

// Log all HTTP requests
if (config.logging.http) {
  logger.info("Logging HTTP requests");

  app.use((req, res, next) => {
    logger.http(`${req.method} ${req.url} ${JSON.stringify(req.body)}`, {
      request: {
        query: req.query,
        params: req.params,
        body: req.body,
        headers: req.headers,
      },
    });
    next();
  });
}

app.use("/static", express.static(path.join(__dirname, "..", "static")));

app.get("/", async (req: Request, res: Response) => {
  res.send(`
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>World-Wide-Lab</title>
    </head>
    <body>
      World-Wide-Lab is running 🌐🧑‍🔬👩‍🔬👨‍🔬 <br />
      <ul>
        <li>The latest API can be found under <a href="/v1/">/v1/</a></li>
      </ul>
    </body>
  </html>
  `);
});

// Mount the API router under v1
app.use("/v1", api);

// Mount the api-docs
if (config.apiDocs.enabled) {
  app.use("/api-docs", apiDocs);
}

// Use adminJS
if (config.admin.enabled) {
  // Make protected API routes available in Admin UI
  adminRouter.use("/wwl/", routerProtectedWithoutAuthentication);

  app.use(admin.options.rootPath, adminRouter);
}

export default app;
