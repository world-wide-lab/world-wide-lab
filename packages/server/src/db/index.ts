import { Sequelize, Options as SequelizeOptions } from "sequelize";

import { defineModels } from "./models";
import config from "../config";
import { log_sql } from "../logger";

const url = config.database.url;
const options: SequelizeOptions = {
  logging: log_sql,
};

// There is a bug in handling filepaths with spaces when using sqlite via
// a URL string. Therefore we manually pass it the database URL for sqlite.
const isSqlite = url.startsWith("sqlite:");
if (isSqlite) {
  // Manually parse URL instead and add it to options
  options.dialect = "sqlite";
  options.storage = url.replace("sqlite://", "").replace("sqlite:", "");
}
const sequelize = isSqlite
  ? new Sequelize(options)
  : new Sequelize(url, options);

defineModels(sequelize);

export default sequelize;
