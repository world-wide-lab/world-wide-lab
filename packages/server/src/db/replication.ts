import type { Model, ModelStatic } from "sequelize";
import config from "../config";
import sequelize from "../db";
import { logger } from "../logger";
import { getLatestMigration } from "./migrate";

const defaultRequestHeaders = {
  "User-Agent": `WWL Replication / ${config.version}`,
};

class UnknownTableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnknownTableError";
  }
}

function findModelByTableName(tableName: string) {
  const model = Object.values(sequelize.models).filter(
    (model) => model.tableName === tableName,
  )[0];

  if (!model) {
    throw new UnknownTableError(`Table "${tableName}" not found`);
  }
  return model;
}

function getNonPrimaryKeyColumns(model: ModelStatic<Model>): string[] {
  const attributes = model.getAttributes();
  const nonPrimaryKeyColumns: string[] = [];
  for (const attr in attributes) {
    if (!attributes[attr].primaryKey) {
      nonPrimaryKeyColumns.push(attr);
    }
  }
  return nonPrimaryKeyColumns;
}

// Import table data into the database
async function importTableData(tableName: string, tableData: any[]) {
  const model = findModelByTableName(tableName);

  console.log(`Importing ${tableData.length} rows into ${tableName}`);

  await model.bulkCreate(tableData, {
    updateOnDuplicate: getNonPrimaryKeyColumns(model),
  });

  model.getAttributes();
}

// Retrieve data from the source
async function fetchTableDataFromSource(
  tableName: string,
  limit: number,
  offset: number,
) {
  logger.info(`Fetching ${tableName} (L:${limit}; O:${offset})`);

  const model = findModelByTableName(tableName);
  const lastUpdated = (await model.max("updatedAt")) as Date;

  const search = new URLSearchParams({
    limit: String(limit),
    offset: String(offset),
    ...(lastUpdated && { updated_after: lastUpdated.toISOString() }),
  }).toString();
  const url = `${config.replication.source}/v1/replication/source/get-table/${tableName}/?${search}`;

  logger.verbose(url);
  const result = await fetch(url, {
    method: "get",
    headers: new Headers({
      ...defaultRequestHeaders,
      Authorization: `Bearer ${config.replication.sourceApiKey}`,
    }),
  });

  const tableData = await result.json();

  // Check whether tableData is an array, else throw an error since sth is wrong
  if (Array.isArray(tableData)) {
    return tableData;
  }
  const statusCode = result.status;
  const hasSourceErrorMessage =
    typeof tableData === "object" && tableData.error;
  if (hasSourceErrorMessage || statusCode !== 200) {
    throw new Error(
      `Replication source reported error when fetching data (${statusCode}). Message: '${
        hasSourceErrorMessage ? tableData.error : ""
      }'.`,
    );
  }
  throw new Error(
    `Error fetching data from source. Table data is not an array: ${tableData}.`,
  );
}

// Get a version identifier for the database
// Curerntly this corresponds to the name of the latest applied migration
// This version is used to check whether the source and destination databases are compatible
async function getDbVersion(): Promise<string> {
  return await getLatestMigration(false);
}

// Check whether the source and destination databases are compatible with each other
async function verifyDatabaseVersion() {
  const result = await fetch(`${config.replication.source}/v1/info`, {
    method: "get",
    headers: new Headers({ ...defaultRequestHeaders }),
  });
  const sourceInfo = await result.json();
  const sourceDbVersion = sourceInfo.db_version;
  const destinationDbVersion = await getDbVersion();

  if (sourceDbVersion !== destinationDbVersion) {
    throw new Error(
      `Database version mismatch. Source: ${sourceDbVersion}, Destination: ${destinationDbVersion} (this machine).`,
    );
  }
}

async function replicateTable(tableName: string) {
  const limit = config.replication.chunkSize;
  let offset = 0;
  let rowCount = limit;
  while (rowCount == limit) {
    const tableData = await fetchTableDataFromSource(tableName, limit, offset);
    rowCount = tableData.length;

    await importTableData(tableName, tableData);

    offset += limit;
  }
}

// Perform a full replication update across all supported tables
async function runReplication() {
  logger.info("Starting replication");

  // Check whether both databases are compatible
  await verifyDatabaseVersion();
  logger.info("Database versions OK");

  const tablesToReplicate = [
    // sequelize.models.Study.tableName,
    // sequelize.models.Participant.tableName,
    // sequelize.models.Session.tableName,
    // sequelize.models.Responses.tableName,
    "wwl_studies",
    "wwl_participants",
    "wwl_sessions",
    "wwl_responses",
  ];

  // Replicate each database table one by one
  for (const tableName of tablesToReplicate) {
    await replicateTable(tableName);
  }

  logger.info("Finished replication.");
}

export {
  UnknownTableError,
  findModelByTableName,
  runReplication,
  getDbVersion,
};
