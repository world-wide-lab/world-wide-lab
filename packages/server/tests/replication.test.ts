// Set up fake environment variables
import "./setup_env";

import { URL } from "url";
import { Op } from "sequelize";
import request from "supertest";
import app from "../src/app";
import config from "../src/config";
import sequelize from "../src/db";
import generateExampleData from "../src/db/exampleData";
import { up } from "../src/db/migrate";
import { UnknownTableError, findModelByTableName } from "../src/db/replication";

const endpoint = request(app);

const API_KEY = process.env.DEFAULT_API_KEY;

const updatedAfter = new Date("2000-01-01T00:00:00Z").toISOString();

describe("Replication", () => {
  beforeAll(async () => {
    // Initialize Database
    await up();
    await sequelize.sync();
  });

  describe("Setup (Example Data)", () => {
    it("should create example data", async () => {
      // Generate example data twice
      // TODO: Mock dates to allow for testing the function of updated_after
      await generateExampleData(sequelize);
      await generateExampleData(sequelize, "example2");

      const nStudies = await sequelize.models.Study.count();
      expect(nStudies).toBe(1 * 2);

      const nParticipants = await sequelize.models.Participant.count();
      expect(nParticipants).toBe(4 * 2);

      const nSessions = await sequelize.models.Session.count();
      expect(nSessions).toBe(8 * 2);

      const nResponses = await sequelize.models.Response.count();
      expect(nResponses).toBe(20 * 2);
    });
  });

  describe("Function: findModelByTableName", () => {
    it("should return the model for a valid table name", () => {
      const model = findModelByTableName("wwl_participants");
      expect(model).toBe(sequelize.models.Participant);
    });

    it("should throw an UnknownTableError for an invalid table name", () => {
      expect(() => findModelByTableName("invalidTable")).toThrow(
        UnknownTableError,
      );
    });
  });

  describe("API of Replication Source", () => {
    beforeAll(() => {
      config.replication.role = "source";
    });

    it("should download a raw list of studies", async () => {
      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_studies?limit=2&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(2);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should retrieve a list of participants", async () => {
      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_participants?limit=5&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(5);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should download a raw list of sessions", async () => {
      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_sessions?limit=12&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(12);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should download a raw list of sessions, honoring offsets", async () => {
      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_sessions?limit=12&offset=8&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(8);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should download a raw list of responses, honoring offsets, without updated_after", async () => {
      const response = await endpoint
        .get(
          "/v1/replication/source/get-table/wwl_responses?limit=25&offset=22",
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(18);
    });

    it("should download a raw list of responses, with multiple chunks", async () => {
      config.database.chunkSize = 2;
      const response = await endpoint
        .get("/v1/replication/source/get-table/wwl_responses?limit=25&offset=3")
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(25);
    });

    it("should download a raw list of responses", async () => {
      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_responses?limit=32&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(32);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should support proper chunking of data", async () => {
      config.database.chunkSize = 2;

      const response = await endpoint
        .get("/v1/replication/source/get-table/wwl_responses?limit=6&offset=12")
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(200);
      expect(response.body.length).toBe(6);
      expect(Object.keys(response.body[0])).toMatchSnapshot();
    });

    it("should fail when not authenticated", async () => {
      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_studies?limit=2&updated_after=${updatedAfter}`,
        )
        .send();

      expect(response.status).toBe(401);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when not configured to act as replication source", async () => {
      config.replication.role = "destination";

      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_studies?limit=2&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(418);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when not configured to act as replication source", async () => {
      config.replication.role = null;

      const response = await endpoint
        .get(
          `/v1/replication/source/get-table/wwl_responses?limit=32&updated_after=${updatedAfter}`,
        )
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(418);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe("API of Replication Destination", () => {
    beforeAll(async () => {
      const REPLICATION_SOURCE_URL = "https://wwl-test";

      // Overwrite configuration
      config.replication.role = "destination";
      config.replication.source = REPLICATION_SOURCE_URL;
      config.replication.sourceApiKey = "non-existent-key";
      config.replication.chunkSize = 2;

      const infoRepsonse = await endpoint.get("/v1/info").send();
      const infoData = infoRepsonse.body;

      // Set up Mocks
      const MOCK_RESPONSES = {
        "/v1/info": {
          GET: () => {
            // Return own info object, so that it always matches the version
            return infoData;
          },
        },
        "/v1/replication/source/get-table/wwl_responses/?": {
          GET: [
            {
              responseId: 1,
              createdAt: "2024-02-13T22:16:56.539Z",
              updatedAt: "2024-02-13T22:16:56.539Z",
              name: "trial-1",
              payload: {
                response: "Response #1",
              },
              sessionId: "8c8facd9-050e-4643-abe5-ed563e8da70a",
            },
            {
              responseId: 2,
              createdAt: "2024-02-13T22:16:56.540Z",
              updatedAt: "2024-02-13T22:16:56.540Z",
              name: "trial-2",
              payload: {
                response: "Response #2",
              },
              sessionId: "8c8facd9-050e-4643-abe5-ed563e8da70a",
            },
            {
              responseId: 3,
              createdAt: "2024-02-13T22:16:56.540Z",
              updatedAt: "2024-02-13T22:16:56.540Z",
              name: "trial-3",
              payload: {
                response: "Response #3",
              },
              sessionId: "8c8facd9-050e-4643-abe5-ed563e8da70a",
            },
            {
              responseId: 4,
              createdAt: "2024-02-13T22:16:56.544Z",
              updatedAt: "2024-02-13T22:16:56.544Z",
              name: "trial-1",
              payload: {
                response: "Response #1",
              },
              sessionId: "d6533e79-ea6b-4b04-a142-946f5e861b43",
            },
            {
              responseId: 5,
              createdAt: "2024-02-13T22:16:56.544Z",
              updatedAt: "2024-02-13T22:16:56.544Z",
              name: "trial-2",
              payload: {
                response: "Response #2",
              },
              sessionId: "d6533e79-ea6b-4b04-a142-946f5e861b43",
            },
          ],
        },
        "/v1/replication/source/get-table/wwl_studies/?": {
          GET: [
            {
              studyId: "replication-test",
              createdAt: "2024-02-13T22:16:56.533Z",
              updatedAt: "2024-02-13T22:16:56.533Z",
              privateInfo: null,
              publicInfo: null,
              deletionProtection: true,
            },
          ],
        },
        "/v1/replication/source/get-table/wwl_participants/?": {
          GET: [
            {
              participantId: "38c10788-6062-43d3-a2e7-cfa929ddaf57",
              createdAt: "2024-02-09T22:16:56.536Z",
              updatedAt: "2024-02-13T22:16:56.536Z",
              privateInfo: {
                description: "Example User #1",
              },
              publicInfo: null,
            },
            {
              participantId: "6b5112d6-93ba-471f-9d76-7deec118ab89",
              createdAt: "2024-02-10T22:16:56.542Z",
              updatedAt: "2024-02-13T22:16:56.542Z",
              privateInfo: {
                description: "Example User #2",
              },
              publicInfo: null,
            },
          ],
        },
        "/v1/replication/source/get-table/wwl_sessions/?": {
          GET: [
            {
              sessionId: "8c8facd9-050e-4643-abe5-ed563e8da70a",
              createdAt: "2024-02-09T22:16:56.536Z",
              updatedAt: "2024-02-13T22:16:56.538Z",
              privateInfo: null,
              publicInfo: null,
              finished: true,
              participantId: "38c10788-6062-43d3-a2e7-cfa929ddaf57",
              studyId: "replication-test",
              metadata: null,
            },
            {
              sessionId: "d6533e79-ea6b-4b04-a142-946f5e861b43",
              createdAt: "2024-02-10T22:16:56.542Z",
              updatedAt: "2024-02-13T22:16:56.543Z",
              privateInfo: null,
              publicInfo: null,
              finished: true,
              participantId: "6b5112d6-93ba-471f-9d76-7deec118ab89",
              studyId: "replication-test",
              metadata: null,
            },
            {
              sessionId: "0ec73e4e-6aea-42cc-bfee-8f700bfa1bcd",
              createdAt: "2024-02-10T22:16:56.542Z",
              updatedAt: "2024-02-10T22:16:56.542Z",
              privateInfo: null,
              publicInfo: null,
              finished: true,
              participantId: "38c10788-6062-43d3-a2e7-cfa929ddaf57",
              studyId: "replication-test",
              metadata: null,
            },
            {
              sessionId: "be218a6f-b3f7-4b01-84e0-c26a8782dd0b",
              createdAt: "2024-02-10T22:16:56.542Z",
              updatedAt: "2024-02-13T22:16:56.543Z",
              privateInfo: null,
              publicInfo: null,
              finished: true,
              participantId: "6b5112d6-93ba-471f-9d76-7deec118ab89",
              studyId: "replication-test",
              metadata: null,
            },
          ],
        },
      };
      // @ts-ignore (typescript doesn't recognize the mock function)
      global.fetch = jest.fn((fetchUrl: string, fetchOptions) => {
        // Mock just one option of using fetch
        const method = fetchOptions?.method?.toUpperCase() || "GET";
        const endpoint = fetchUrl.replace(REPLICATION_SOURCE_URL, "");

        const matchingEndpoints = Object.keys(MOCK_RESPONSES).filter((k) =>
          endpoint.startsWith(k),
        );
        if (matchingEndpoints.length != 1) {
          throw new Error(
            `N = ${matchingEndpoints.length} mock responses are matching the provided endpoint: ${endpoint}.`,
          );
        }

        // @ts-ignore
        const matchingRespone = MOCK_RESPONSES[matchingEndpoints[0]][method];
        if (matchingRespone === undefined) {
          throw new Error(`No mock response for "${endpoint}" "${method}"`);
        }

        let data: Array<object> | object;
        if (typeof matchingRespone == "function") {
          // Function => execute it
          data = matchingRespone({ endpoint });
        } else if (Array.isArray(matchingRespone)) {
          // Array => get slices via limit and offset
          const url = new URL(endpoint, "http://localhost");
          const limit = parseInt(url.searchParams.get("limit") as string);
          const offset = parseInt(url.searchParams.get("offset") as string);

          data = matchingRespone.slice(offset, offset + limit);
        } else {
          // Else just return as is
          data = matchingRespone;
        }

        return Promise.resolve({
          json: () => Promise.resolve(data),
        });
      }) as jest.MockedFunction<typeof fetch>;
    });

    it("should work as intended", async () => {
      const response = await endpoint
        .get(`/v1/replication/destination/update`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      // Perform multiple mocked fetch requests here as part of the update

      // Response should be correct
      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      // New data should be inserted
      const nStudies = await sequelize.models.Study.count({
        where: {
          studyId: "replication-test",
        },
      });
      expect(nStudies).toBe(1);
      const sessions = await sequelize.models.Session.findAll({
        where: {
          studyId: "replication-test",
        },
      });
      expect(sessions.length).toBe(4);
      // @ts-ignore
      const participantIds = sessions.map((s) => s.participantId);
      // @ts-ignore
      const sessionIds = sessions.map((s) => s.sessionId);

      const nParticipants = await sequelize.models.Participant.count({
        where: {
          participantId: {
            [Op.in]: participantIds,
          },
        },
      });
      expect(nParticipants).toBe(2);

      const nResponses = await sequelize.models.Response.count({
        where: {
          sessionId: {
            [Op.in]: sessionIds,
          },
        },
      });
      expect(nResponses).toBe(5);
    });

    it("should fail when not configured to act as destination source", async () => {
      config.replication.role = null;

      const response = await endpoint
        .get(`/v1/replication/destination/update`)
        .set("Authorization", `Bearer ${API_KEY}`)
        .send();

      expect(response.status).toBe(418);
      expect(response.body).toMatchSnapshot();
    });

    it("should fail when not authenticated", async () => {
      const response = await endpoint
        .get(`/v1/replication/destination/update`)
        .send();

      expect(response.status).toBe(401);
      expect(response.body).toMatchSnapshot();
    });
  });
});
