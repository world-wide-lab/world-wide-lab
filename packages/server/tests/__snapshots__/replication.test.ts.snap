// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Replication API of Replication Destination should fail when not authenticated 1`] = `
{
  "error": "Authentication via API Key required",
}
`;

exports[`Replication API of Replication Destination should fail when not configured to act as destination source 1`] = `
{
  "error": "Serving as a replication destination is not enabled. Set REPLICATION_ROLE to 'destination' to enable this feature.",
}
`;

exports[`Replication API of Replication Destination should work as intended 1`] = `
{
  "message": "Success!",
}
`;

exports[`Replication API of Replication Source should download a raw list of responses 1`] = `
[
  "responseId",
  "createdAt",
  "updatedAt",
  "name",
  "payload",
  "sessionId",
]
`;

exports[`Replication API of Replication Source should download a raw list of sessions 1`] = `
[
  "sessionId",
  "createdAt",
  "updatedAt",
  "privateInfo",
  "publicInfo",
  "finished",
  "participantId",
  "studyId",
  "metadata",
]
`;

exports[`Replication API of Replication Source should download a raw list of sessions, honoring offsets 1`] = `
[
  "sessionId",
  "createdAt",
  "updatedAt",
  "privateInfo",
  "publicInfo",
  "finished",
  "participantId",
  "studyId",
  "metadata",
]
`;

exports[`Replication API of Replication Source should download a raw list of studies 1`] = `
[
  "studyId",
  "createdAt",
  "updatedAt",
  "privateInfo",
  "publicInfo",
  "deletionProtection",
]
`;

exports[`Replication API of Replication Source should fail when not authenticated 1`] = `
{
  "error": "Authentication via API Key required",
}
`;

exports[`Replication API of Replication Source should fail when not configured to act as replication source 1`] = `
{
  "error": "Serving as a replication source is not enabled. Set REPLICATION_ROLE to 'source' to enable this feature.",
}
`;

exports[`Replication API of Replication Source should fail when not configured to act as replication source 2`] = `
{
  "error": "Serving as a replication source is not enabled. Set REPLICATION_ROLE to 'source' to enable this feature.",
}
`;

exports[`Replication API of Replication Source should retrieve a list of participants 1`] = `
[
  "participantId",
  "createdAt",
  "updatedAt",
  "privateInfo",
  "publicInfo",
]
`;

exports[`Replication API of Replication Source should support proper chunking of data 1`] = `
[
  "responseId",
  "createdAt",
  "updatedAt",
  "name",
  "payload",
  "sessionId",
]
`;
